src/fuzz_introspector/frontends/datatypes.py

```diff
@@ -14,13 +14,78 @@
 #
 ################################################################################
 
-from typing import Any, Optional
+from typing import Any, Optional, Generic, TypeVar
 
+from tree_sitter import Language, Parser
+import tree_sitter_c
+import tree_sitter_cpp
+import tree_sitter_go
+import tree_sitter_java
+import tree_sitter_rust
 
-class Project():
+import logging
+
+logger = logging.getLogger(name=__name__)
+
+T = TypeVar('T', bound='SourceCodeFile')
+
+
+class SourceCodeFile():
+    """Class for holding file-specific information."""
+    LANGUAGE: dict[str, Language] = {
+        'c': Language(tree_sitter_c.language()),
+        'cpp': Language(tree_sitter_cpp.language()),
+        'c++': Language(tree_sitter_cpp.language()),
+        'go': Language(tree_sitter_go.language()),
+        'jvm': Language(tree_sitter_java.language()),
+        'rust': Language(tree_sitter_rust.language()),
+    }
+
+    def __init__(self,
+                 language: str,
+                 source_file: str,
+                 entrypoint: str = '',
+                 source_content: Optional[bytes] = None):
+        logger.info('Processing %s' % source_file)
+
+        self.root = None
+        self.source_file = source_file
+        self.language = language
+        self.entrypoint = entrypoint
+        self.tree_sitter_lang = self.LANGUAGE.get(language)
+        self.parser = Parser(self.tree_sitter_lang)
+
+        if source_content:
+            self.source_content = source_content
+        else:
+            with open(self.source_file, 'rb') as f:
+                self.source_content = f.read()
+
+        # Initialization ruotines
+        self.load_tree()
+
+        # Language specific process
+        self.language_specific_process()
+
+    def load_tree(self):
+        """Load the the source code into a treesitter tree, and set
+        the root node."""
+        if not self.root:
+            self.root = self.parser.parse(self.source_content).root_node
+
+    def language_specific_process(self):
+        """Dummy function to perform some specific processes in subclasses."""
```

Comment:
am not sure if we should use `dummy` here, or just give documentation as how classes inherit from here should use the function